### PR TITLE
ESQL: Add more time span units

### PR DESCRIPTION
**Pull request by bpintea**
This adds ~`nanosecond`, `microsecond` and~ `quarter` to the set of supported time spans. It also adds a few standard and common abbreviations to some existing ones.

Edit: not adding sub-millisecond units, since these aren't currently supported (resolution currently support millis only).

### CODE REVIEW

File: docs/reference/esql/esql-syntax.asciidoc [hunk 1]
```diff
@@ -160,14 +160,17 @@ Datetime intervals and timespans can be expressed using timespan literals.
...
-* `hour`/`hours`
-* `day`/`days`
+* `nanosecond`/`nanoseconds`/`ns`
+* `microsecond`/`microseconds`/`us`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`s`
+* `minute`/`minutes`/`min`
+* `hour`/`hours`/`h`
+* `day`/`days`/`d`
 * `week`/`weeks`
```
Comment (costin) [inline docs/reference/esql/esql-syntax.asciidoc:170]: How about `w` as a shorthand for weeks? There doesn't seem to be any ambiguity (especially not around the near units).

File: docs/reference/esql/esql-syntax.asciidoc [hunk 2]
```diff
@@ -160,14 +160,17 @@ Datetime intervals and timespans can be expressed using timespan literals.
...
-* `month`/`months`
-* `year`/`years`
+* `month`/`months`/`mo`
+* `quarter`/`quarters`
+* `year`/`years`/`yr`
```
Comment (costin) [inline docs/reference/esql/esql-syntax.asciidoc:173]: How about `y` : `now() - 1y`? Doesn't seem to be any other y unit near it.

**Review by costin — APPROVED** [review]
LGTM

File: docs/reference/esql/esql-syntax.asciidoc (hunk 2 above)
Comment (alex-spies) [reply docs/reference/esql/esql-syntax.asciidoc:173]: ++

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java [hunk 3]
```diff
@@ -373,32 +373,47 @@ public void testUnquotedIdentifiers() {
     public void testDurationLiterals() {
         int value = randomInt(Integer.MAX_VALUE);
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java:374]: Super nit: Maybe we should use a random long instead to better cover the relevant range (esp. for nanos).

**Review by alex-spies — COMMENTED** [review]
Nice addition! But I think we need to add more tests.
- `AbstractFunctionTestCase.randomLiteral()` never emits `time_duration`s in units smaller than millis. The same applies to `TestCaseSupplier.timeDurationCases`.
- None of our csv tests use the new units, so there's no "proof" that they actually work end-to-end, resp. nothing guarding against regressions. At least, I think we should have some date arithmetic tests for `datetime +/- x` where `x` is given in nanos, mus, and quarters.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java (hunk 3 above)
Comment (bpintea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java]: IIRC some of the static Duration/Period methods fail with too large values and this was the reason to use an int here.

File: docs/reference/esql/esql-syntax.asciidoc (hunk 1 above)
Comment (bpintea) [reply docs/reference/esql/esql-syntax.asciidoc]: I've added a `q` for quarter too, btw.

**Comment by bpintea** [comment]
> AbstractFunctionTestCase.randomLiteral() never emits time_durations in units smaller than millis. The same applies to TestCaseSupplier.timeDurationCases.

And they should stay the same, since it wouldn't make a difference, as we don't support nanos. Which means it makes no sense adding them a.t.p. -- I've removed them. 🙏 

> At least, I think we should have some date arithmetic tests for datetime +/- x where x is given in nanos, mus, and quarters.

I've added some csv-spec tests with quarter and the new abbreviations. I think the unit tests would have been relevant for parsing these units, but more tests is better than less tests.
(Btw, csv-spec tests with sub-millisecond units would have worked, but these wouldn't have made a difference to the outcome.)

File: docs/reference/esql/esql-syntax.asciidoc [hunk 4]
```diff
@@ -160,14 +160,15 @@ Datetime intervals and timespans can be expressed using timespan literals.
...
-* `year`/`years`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`s`
+* `minute`/`minutes`/`min`
+* `hour`/`hours`/`h`
+* `day`/`days`/`d`
+* `week`/`weeks`/`w`
+* `month`/`months`/`mo`
+* `quarter`/`quarters`/`q`
+* `year`/`years`/`yr`/`y`
```
Comment (alex-spies) [inline docs/reference/esql/esql-syntax.asciidoc:171]: - Are we sure we want to deviate from the [standard abbreviations](https://www.elastic.co/guide/en/elasticsearch/reference/8.13/common-options.html#date-math) used in ES? Applies to `mo` and `min`. In SQL, apparently there is yet another [set of abbreviations](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-add); not sure if we should follow suit, but having 3 different sets of abbreviations will likely be confusing.
- Are we sure that the time period `min` and the aggregation function `min` will never be ambiguous when parsing?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java [hunk 5]
```diff
@@ -234,18 +234,20 @@ public static DataType commonType(DataType left, DataType right) {
         return DataTypeConverter.commonType(left, right);
     }
 
+    // generally supporting abbreviations from https://en.wikipedia.org/wiki/Unit_of_time
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java:237]: Then we should also add smaller time units, like one [Jiffy](https://en.wikipedia.org/wiki/Unit_of_time#Scientific) and the Planck time. /s

File: docs/reference/esql/esql-syntax.asciidoc [hunk 6]
```diff
@@ -160,14 +160,15 @@ Datetime intervals and timespans can be expressed using timespan literals.
...
-* `millisecond`/`milliseconds`
-* `second`/`seconds`
-* `minute`/`minutes`
-* `hour`/`hours`
-* `day`/`days`
-* `week`/`weeks`
-* `month`/`months`
-* `year`/`years`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`s`
```
Comment (alex-spies) [inline docs/reference/esql/esql-syntax.asciidoc:164]: Missing `sec`
```suggestion
* `second`/`seconds`/`sec`/`s`
```

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java [hunk 7]
```diff
@@ -380,37 +380,51 @@ public void testDurationLiterals() {
         assertEquals(l(Duration.ZERO, TIME_DURATION), whereExpression("0 second"));
         assertEquals(l(Duration.ofSeconds(value), TIME_DURATION), whereExpression(value + "second"));
         assertEquals(l(Duration.ofSeconds(value), TIME_DURATION), whereExpression(value + " seconds"));
+        assertEquals(l(Duration.ofSeconds(value), TIME_DURATION), whereExpression(value + " s"));
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java:383]: Missing `sec`.

File: docs/reference/esql/esql-syntax.asciidoc (hunk 4 above)
Comment (bpintea) [reply docs/reference/esql/esql-syntax.asciidoc:171]: > In SQL, apparently there is yet another [set of abbreviations](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-add); not sure if we should follow suit, but having 3 different sets of abbreviations will likely be confusing.

Interestingly that you've made this comparison too. I've recently added [some clarifications](https://github.com/elastic/elasticsearch/pull/108301) about the non-overlapping of this set with function parameters.

> Are we sure that the time period min and the aggregation function min will never be ambiguous when parsing?

We have the tests to keep us in check. But I also hope we won't need to change the grammar in a way that there might be conflicts.

File: docs/reference/esql/esql-syntax.asciidoc (hunk 6 above)
Comment (bpintea) [reply docs/reference/esql/esql-syntax.asciidoc:164]: Yeh... I guess if we opened the gates, "sec" can be there too. I'll add it.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java (hunk 7 above)
Comment (bpintea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java:383]: Thx!

**Comment by bpintea** [comment]
Thx Costin, Alex.

File: docs/reference/esql/esql-syntax.asciidoc [hunk 8]
```diff
@@ -160,14 +160,15 @@ Datetime intervals and timespans can be expressed using timespan literals.
...
-* `year`/`years`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`sec`/`s`
+* `minute`/`minutes`/`min`
+* `hour`/`hours`/`h`
+* `day`/`days`/`d`
+* `week`/`weeks`/`w`
+* `month`/`months`/`mo`
+* `quarter`/`quarters`/`q`
+* `year`/`years`/`yr`/`y`
```
Comment (drewdaemon) [inline docs/reference/esql/esql-syntax.asciidoc:171]: Looks like we missed `microsecond` and `nanosecond` cc @bpintea

Comment (bpintea) [reply docs/reference/esql/esql-syntax.asciidoc:171]: I've updated the PR description. We don't yet support sub-millisecond values.

**Pull request merged**